module/mailnews/mailread.c

```diff
@@ -353,3 +404,174 @@ void UserMailReply(int msg_num, Bool reply_all)
    MakeReplySubject(reply->subject, MAX_SUBJECT);
    MailSendReply(hReadMailDlg, reply);
 }
+
+void OnColumnClick(LPNMLISTVIEW pLVInfo)
+{
+   int nSortColumn = pLVInfo->iSubItem;
+   bool bSortAscending = TRUE;
+
+   if (nSortColumn == currentSortColumn)
+   {
+      // Toggle sort direction if the same column is clicked again
+      currentSortAscending = !currentSortAscending;
+   }
+   else
+   {
+      // Set ascending order for new column
+      currentSortColumn = nSortColumn;
+      currentSortAscending = TRUE;
+   }
+
+   LPARAM lParamSort = (currentSortAscending ? 1 : -1) * (currentSortColumn + 1);
+
+   if (hbmUpArrow != NULL && hbmDownArrow != NULL)
+   {
+      ListView_SetHeaderSortImage(pLVInfo->hdr.hwndFrom, currentSortColumn, currentSortAscending);
+   }
+
+   // Sort list
+   ListView_SortItemsEx(pLVInfo->hdr.hwndFrom, CompareListItems, lParamSort);
+}
+
+/****************************************************************************/
+/*
+ * CompareListItems: Comparison function for sorting items in the news dialog
+ * list view.
+ */
+int CALLBACK CompareListItems(LPARAM lParam1, LPARAM lParam2, LPARAM lParamSort)
+{
+   bool bSortAscending = (lParamSort > 0);
+   int nColumn = abs(lParamSort) - 1;
+   char szBuf1[MAX_SUBJECT], szBuf2[MAX_SUBJECT];
+   int time1 = 0;
+   int time2 = 0;
+
+   HWND hListView = GetDlgItem(hReadMailDlg, IDC_MAILLIST);
+
+   // Extract the msg_num values from the list view items
+   ListView_GetItemText(hListView, lParam1, COL_ORDER, szBuf1, sizeof(szBuf1));
+   ListView_GetItemText(hListView, lParam2, COL_ORDER, szBuf2, sizeof(szBuf2));
+   int msg_num1 = atoi(szBuf1);
+   int msg_num2 = atoi(szBuf2);
+
+   ListView_GetItemText(hListView, lParam1, nColumn, szBuf1, sizeof(szBuf1));
+   ListView_GetItemText(hListView, lParam2, nColumn, szBuf2, sizeof(szBuf2));
+
+   switch (nColumn)
+   {
+   case COL_ORDER:
+      return bSortAscending ? atoi(szBuf1) - atoi(szBuf2) : atoi(szBuf2) - atoi(szBuf1);
+   case COL_FROM:
+      return bSortAscending ? stricmp(szBuf1, szBuf2) : stricmp(szBuf2, szBuf1);
+   case COL_SUBJECT:
+      return bSortAscending ? stricmp(szBuf1, szBuf2) : stricmp(szBuf2, szBuf1);
+   case COL_TIME:
+      time1 = mail_date_map[msg_num1];
+      time2 = mail_date_map[msg_num2];
+
+      return bSortAscending ? (time1 - time2) : (time2 - time1);
+   default:
+      debug(("Unhandled column (column #%d) in CompareListItems\n", nColumn));
+      return 0;
+   }
+}
+
+void ListView_SetHeaderSortImage(HWND hListView, int sortedColumn, BOOL sortAscending)
```

Comment:
I'm super confused.. there's also a function with the same name declared in the header, so that it's visible everywhere?

- Only put function signatures in headers if they're shared.  Otherwise declare them static in the .c file.
- If the functions are the same, have only one copy.
- Otherwise, name them differently so as not to confuse the reader.